en: add doc for deploying multiple tidb operator (#1057)
pingcap/docs-tidb-operator: en/canary-deployment-tidb-operator.md

---
title: Canary Deployment of TiDB Operator

> summary: Learn how to canary upgrade TiDB Operator and how to deploy multiple TiDB Operator to manage different TiDB clusters.

Reviewer: ```suggestion
summary: Learn how to perform a canary upgrade to TiDB Operator and how to deploy multiple TiDB Operator to manage different TiDB clusters.
```

---

# Canary Deployment of TiDB Operator

> This document describes how to canary upgrade TiDB Operator and how to deploy multiple TiDB Operator to manage different TiDB clusters.

Reviewer: ```suggestion
This document describes how to perform a canary upgrade to TiDB Operator and how to deploy multiple TiDB Operator to manage different TiDB clusters.
```

## Scenarios

> - Canary upgrade TiDB Operator.

Reviewer: ditto

Author: fixed in a832dba

> - Deploy multiple TiDB Operator to manage different TiDB clusters.

Reviewer: Can we say `multiple sets of TiDB Operator` or `multiple TiDB Operator sets` instead of `multiple TiDB Operator`?

Author: Thanks! I've been troubled by the plural forms for a while ... `multiple sets of TiDB Operator` SGTM

Author: fixed in a832dba

    > **Note:**
    >
    > - Currently, you can only deploy multiple `tidb-controller-manager`s and `tidb-scheduler`s.
    > - Only one Admission Webhook is required.

>     > - If you have deployed multiple TiDB Operator and only some of them enable [Advanced StatefulSet](advanced-statefulset.md), the same TidbCluster Custom Resource (CR) cannot be switched among these TiDB Operator.

Reviewer: `TiDB Operator` here looks like a countable noun but it cannot be suffixed with `s` to indicate its plural form, which to me is a bit strange.

## Related parameters

To support canary deployment, the following parameters are added to the `values.yaml` file in the `tidb-operator` chart:

- `appendReleaseSuffix`

>     If this parameter is set to `true`, when you deploy TiDB Operator, TiDB Operator adds automatically a prefix (`-{{ .Release.Name }}`) to the name of resources related to `tidb-controller-manager` and `tidb-scheduler`.

Reviewer: ```suggestion
    If this parameter is set to `true`, when you deploy TiDB Operator, TiDB Operator automatically adds a prefix (`-{{ .Release.Name }}`) to the name of resources related to `tidb-controller-manager` and `tidb-scheduler`.
```

Reviewer 2: No, not TiDB Operator, but the helm chart.

Reviewer 2: ```suggestion
    If this parameter is set to `true`, when you deploy TiDB Operator, the Helm chart automatically adds a suffix (`-{{ .Release.Name }}`) to the name of resources related to `tidb-controller-manager` and `tidb-scheduler`.
```

    For example, if you execute `helm install canary pingcap/tidb-operator ...`, the name of the `tidb-controller-manager` deployment is `tidb-controller-manager-canary`.

    If you need to deploy multiple TiDB Operator, set this parameter to `true`.

    Default value: `false`.

- `controllerManager.create`

    Controls whether to create `tidb-controller-manager`.

    Default value: `true`.

- `controllerManager.selector`

>     Sets the `-selector` parameter of `tidb-controller-manager`. The parameter is used to filter the CRs controlled by `tidb-controller-manager` according to the CR labels. If multiple selectors exist, the selectors are in `and` relationship.

Reviewer: ```suggestion
    Sets the `-selector` parameter for `tidb-controller-manager`. The parameter is used to filter the CRs controlled by `tidb-controller-manager` according to the CR labels. If multiple selectors exist, the selectors are in `and` relationship.
```

    Default value: `[]` (`tidb-controller-manager` controls all CRs).

    Example:

    ```yaml
    selector:
    - canary-release=v1
    - k1==v1
    - k2!=v2
    ```

- `scheduler.create`

    Controls whether to create `tidb-scheduler`.

    Default value: `true`.

## Deploy multiple TiDB Operator to manage different TiDB clusters

1. Deploy the first TiDB Operator.

    Refer to [Deploy TiDB Operator](deploy-tidb-operator.md) to deploy the first TiDB Operator. Add the following configuration in the `values.yaml`:

    ```yaml
    controllerManager:
      selector:
      - user=dev
    ```

2. Deploy the TiDB cluster.

>     1. Refer to [Configure the TiDB Cluster](configure-a-tidb-cluster.md) to configure the TidbCluster CR, and configure `labels` to match the `selector` set in the last step. For example:

Reviewer: labels or label?

Author: In the yaml below it's `labels`, so I took the liberty to change it to `labels`.
@DanielZhangQD PTAL.

Reviewer 2: `labels` is OK.

        ```yaml
        apiVersion: pingcap.com/v1alpha1
        kind: TidbCluster
        metadata:
          name: basic1
          labels:
            user: dev
        spec:
          ...
        ```

>         If `label` is not set when you deploy the TiDB cluster, you can configure `label` by running the following command:

Reviewer 2: ```suggestion
        If `labels` is not set when you deploy the TiDB cluster, you can configure `labels` by running the following command:
```

        {{< copyable "shell-regular" >}}

        ```shell
        kubectl -n ${namespace} label tidbcluster ${cluster_name} user=dev
        ```

    2. Refer to [Deploy TiDB in General Kubernetes](deploy-on-general-kubernetes.md) to deploy the TiDB cluster. Confirm that each component in the cluster is started normally.

3. Deploy the second TiDB Operator.

    Refer to [Deploy TiDB Operator](deploy-tidb-operator.md) to deploy the second TiDB Operator without `tidb-scheduler`. Add the following configuration in the `values.yaml`:

    ```yaml
    controllerManager:
      selector:
      - user=qa
    appendReleaseSuffix: true
    scheduler:
      create: false
    ```

    > **Note:**
    >
    > If you configure `scheduler.create: true`, a `tidb-scheduler` named `{{ .scheduler.schedulerName }}-{{.Release.Name}}` is created. To use this `tidb-scheduler`, you need to configure `spec.schedulerName` in the TidbCluster CR to the name of this scheduler.

4. Deploy the TiDB cluster.

>     1. Refer to [Configure the TiDB Cluster](configure-a-tidb-cluster.md) to configure the TidbCluster CR, and configure `labels` to match the `selector` set in the last step. For example:

Reviewer: labels or label?

        ```yaml
        apiVersion: pingcap.com/v1alpha1
        kind: TidbCluster
        metadata:
          name: basic2
          labels:
            user: qa
        spec:
          ...
        ```

>         If `label` is not set when you deploy the TiDB cluster, you can configure `label` by running the following command:

Reviewer 2: ```suggestion
        If `labels` is not set when you deploy the TiDB cluster, you can configure `labels` by running the following command:
```

        {{< copyable "shell-regular" >}}

        ```shell
        kubectl -n ${namespace} label tidbcluster ${cluster_name} user=qa
        ```

    2. Refer to [Deploy TiDB in General Kubernetes](deploy-on-general-kubernetes.md) to deploy the TiDB cluster. Confirm that each component in the cluster is started normally.

5. View the logs of the two TiDB Operator, and confirm that each TiDB Operator manages the TiDB cluster that matches the corresponding selectors.

    For example:

    View the log of `tidb-controller-manager` of the first TiDB Operator:

    ```shell
    kubectl -n tidb-admin logs tidb-controller-manager-55b887bdc9-lzdwv
    ```

>     <details>

Reviewer: Using the `<details>` syntax for toggle display might affect searching the toggled words, as you have previously found out.

Author: Yes. I agree with you. The Chinese author chose the toggle display. I assume he believed the output is trivial in this context?

Reviewer 2: Yes, it's OK.
